File: food-backend/src/models/like.model.ts
import { Schema, model, Types } from "mongoose";
import { LikeType } from "../types/like";

const likeSchema = new Schema<LikeType>(
  {
    userId: { type: Types.ObjectId, ref: "User", required: true },
    menuItemId: { type: Types.ObjectId, ref: "MenuItem" },
    restaurantId: { type: Types.ObjectId, ref: "Restaurant" },
  },
  { timestamps: true }
);

// Add a pre-save hook to ensure either menuItemId or restaurantId is present, but not both.
likeSchema.pre("save", function (next) {
  if (!this.menuItemId && !this.restaurantId) {
    return next(new Error("Either menuItemId or restaurantId is required."));
  }
  if (this.menuItemId && this.restaurantId) {
    return next(new Error("Cannot like both a menuItem and a restaurant at the same time."));
  }
  next();
});

export const Like = model<LikeType>("Like", likeSchema);

File: lib/models/like_model.dart
import 'menu_item.dart';
class Like {
  final String id;
  final String userId;
  final MenuItem menuItemId;

  Like({
    required this.id,
    required this.userId,
    required this.menuItemId,
  });

  factory Like.fromJson(Map<String, dynamic> json) => Like(
        id: json['_id'],
        userId: json['userId'],
        menuItemId: MenuItem.fromJson(json['menuItemId']),
      );
}